### ISSUE TITLE
Adapter 7.0.5 crashed wenn ts-Scripte laufen

### ISSUE THREAD

**Issue by JohGrei** [ioBroker/ioBroker.javascript#1325]
**Describe the bug**  
nach Update auf Version 5.0.7 crashed der Adapter beim aufruf von ts-Scripts. Sind diese Scripte nicht aktiv startet er normal.

**To Reproduce**  
2. Click on '...'

**Expected behavior**  

**Screenshots & Logfiles**  
`2023-06-07 06:41:20.245 - info: javascript.0 (1319455) script.js.global.IconsSelector: compiling TypeScript source...

2023-06-07 06:41:20.364 - error: javascript.0 (1319455) Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2023-06-07 06:41:20.365 - error: javascript.0 (1319455) unhandled promise rejection: ts.createNodeArray is not a function
2023-06-07 06:41:20.366 - error: javascript.0 (1319455) TypeError: ts.createNodeArray is not a function
at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:368:138
at Object.visitNode (/opt/iobroker/node_modules/typescript/lib/typescript.js:85924:21)
at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:359:16
at transformation (/opt/iobroker/node_modules/typescript/lib/typescript.js:109613:16)
at transformNodes (/opt/iobroker/node_modules/typescript/lib/typescript.js:109621:72)
at Object.transform (/opt/iobroker/node_modules/typescript/lib/typescript.js:141644:20)
at transformScriptBeforeCompilation (/opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:447:23)
at Immediate. (/opt/iobroker/node_modules/iobroker.javascript/main.js:1123:67)
at processImmediate (node:internal/timers:478:21)
2023-06-07 06:41:20.366 - error: javascript.0 (1319455) ts.createNodeArray is not a function
2023-06-07 06:41:20.422 - info: javascript.0 (1319455) terminating
2023-06-07 06:41:20.424 - warn: javascript.0 (1319455) Terminated (UNCAUGHT_EXCEPTION): Without reason
2023-06-07 06:41:21.044 - error: host.vm-iobroker Caught by controller[0]: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
2023-06-07 06:41:21.074 - error: host.vm-iobroker Caught by controller[1]: TypeError: ts.createNodeArray is not a function
2023-06-07 06:41:21.074 - error: host.vm-iobroker Caught by controller[1]: at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:368:138
2023-06-07 06:41:21.074 - error: host.vm-iobroker Caught by controller[1]: at Object.visitNode (/opt/iobroker/node_modules/typescript/lib/typescript.js:85924:21)
2023-06-07 06:41:21.075 - error: host.vm-iobroker Caught by controller[1]: at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:359:16
2023-06-07 06:41:21.075 - error: host.vm-iobroker Caught by controller[1]: at transformation (/opt/iobroker/node_modules/typescript/lib/typescript.js:109613:16)
2023-06-07 06:41:21.075 - error: host.vm-iobroker Caught by controller[1]: at transformNodes (/opt/iobroker/node_modules/typescript/lib/typescript.js:109621:72)
2023-06-07 06:41:21.076 - error: host.vm-iobroker Caught by controller[1]: at Object.transform (/opt/iobroker/node_modules/typescript/lib/typescript.js:141644:20)
2023-06-07 06:41:21.076 - error: host.vm-iobroker Caught by controller[1]: at transformScriptBeforeCompilation (/opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:447:23)
2023-06-07 06:41:21.076 - error: host.vm-iobroker Caught by controller[1]: at Immediate. (/opt/iobroker/node_modules/iobroker.javascript/main.js:1123:67)
2023-06-07 06:41:21.076 - error: host.vm-iobroker Caught by controller[1]: at processImmediate (node:internal/timers:478:21)
2023-06-07 06:41:21.077 - error: host.vm-iobroker instance system.adapter.javascript.0 terminated with code 6 (UNCAUGHT_EXCEPTION)
2023-06-07 06:41:54.763 - info: javascript.0 (1320607) starting. Version 7.0.5 in /opt/iobroker/node_modules/iobroker.javascript, node: v18.16.0, js-controller: 4.0.24`

**Versions:**  
 - Adapter version: 7.0.5
 - JS-Controller version: 4.0.24
 - Node version: 18.16.0
 - Operating system: linux

**Additional context**

**Comment by klein0r** [comment]
Passiert das mit jedem beliebigen TS? IconsSelector klingt nach NSPanel? 😄

**Comment by JohGrei** [comment]
> Passiert das mit jedem beliebigen TS? IconsSelector klingt nach NSPanel? 😄

ja, NSPanel, hab da zwei ts Scripte am Laufen, andere nicht

**Comment by Garfonso** [comment]
I can second this. Happens with all Typescript scripts. If I disable all of them, the adapter works fine.

**Comment by Harry-58** [comment]
Bei mir das Gleiche. Wenn ich alle TS-Scripte deaktiviere, läuft der Adapter.
Wenn ich ein TS-Script starte, crashed der Adapter sofort.

2023-06-12 07:41:29.149 - info: javascript.0 (26307) script.js.common.Anrufer_Melden: compiling TypeScript source...
2023-06-12 07:41:29.296 - error: javascript.0 (26307) uncaught exception: ts.createNodeArray is not a function
2023-06-12 07:41:29.297 - error: javascript.0 (26307) TypeError: ts.createNodeArray is not a function
at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:368:138
at Object.visitNode (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:85924:21)
at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:359:16
at transformation (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:109613:16)
at transformNodes (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:109621:72)
at Object.transform (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:141644:20)
at transformScriptBeforeCompilation (/opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:447:23)
at prepareScript (/opt/iobroker/node_modules/iobroker.javascript/main.js:2068:39)
at /opt/iobroker/node_modules/iobroker.javascript/main.js:2153:17
at Immediate. (/opt/iobroker/node_modules/iobroker.javascript/main.js:1628:17)
at processImmediate (node:internal/timers:478:21)
2023-06-12 07:41:29.297 - error: javascript.0 (26307) ts.createNodeArray is not a function

2023-06-12 07:41:29.346 - info: javascript.0 (26307) terminating
2023-06-12 07:41:29.349 - warn: javascript.0 (26307) Terminated (UNCAUGHT_EXCEPTION): Without reason
2023-06-12 07:41:30.293 - error: host.iobroker Caught by controller[0]: TypeError: ts.createNodeArray is not a function
2023-06-12 07:41:30.294 - error: host.iobroker Caught by controller[0]: at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:368:138
2023-06-12 07:41:30.294 - error: host.iobroker Caught by controller[0]: at Object.visitNode (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:85924:21)
2023-06-12 07:41:30.294 - error: host.iobroker Caught by controller[0]: at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:359:16
2023-06-12 07:41:30.295 - error: host.iobroker Caught by controller[0]: at transformation (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:109613:16)
2023-06-12 07:41:30.295 - error: host.iobroker Caught by controller[0]: at transformNodes (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:109621:72)
2023-06-12 07:41:30.295 - error: host.iobroker Caught by controller[0]: at Object.transform (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:141644:20)
2023-06-12 07:41:30.295 - error: host.iobroker Caught by controller[0]: at transformScriptBeforeCompilation (/opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:447:23)
2023-06-12 07:41:30.295 - error: host.iobroker Caught by controller[0]: at prepareScript (/opt/iobroker/node_modules/iobroker.javascript/main.js:2068:39)
2023-06-12 07:41:30.295 - error: host.iobroker Caught by controller[0]: at /opt/iobroker/node_modules/iobroker.javascript/main.js:2153:17
2023-06-12 07:41:30.295 - error: host.iobroker Caught by controller[0]: at Immediate. (/opt/iobroker/node_modules/iobroker.javascript/main.js:1628:17)
2023-06-12 07:41:30.296 - error: host.iobroker Caught by controller[0]: at processImmediate (node:internal/timers:478:21)
2023-06-12 07:41:30.296 - error: host.iobroker instance system.adapter.javascript.0 terminated with code 6 (UNCAUGHT_EXCEPTION)
2023-06-12 07:41:30.302 - info: host.iobroker Restart adapter system.adapter.javascript.0 because enabled
2023-06-12 07:42:00.500 - info: host.iobroker instance system.adapter.javascript.0 started with pid 26815
2023-06-12 07:42:04.324 - info: javascript.0 (26815) starting. Version 7.0.5 in /opt/iobroker/node_modules/iobroker.javascript, node: v18.16.0, js-controller: 4.0.24
2023-06-12 07:42:05.085 - info: javascript.0 (26815) requesting all states
2023-06-12 07:42:05.087 - info: javascript.0 (26815) requesting all objects
2023-06-12 07:42:06.248 - info: javascript.0 (26815) received all objects
2023-06-12 07:42:06.387 - info: javascript.0 (26815) received all states
2023-06-12 07:42:12.943 - info: javascript.0 (26815) script.js.common.Anrufer_Melden: compiling TypeScript source...
2023-06-12 07:42:13.087 - error: javascript.0 (26815) uncaught exception: ts.createNodeArray is not a function
2023-06-12 07:42:13.088 - error: javascript.0 (26815) TypeError: ts.createNodeArray is not a function
at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:368:138
at Object.visitNode (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:85924:21)
at /opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:359:16
at transformation (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:109613:16)
at transformNodes (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:109621:72)
at Object.transform (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:141644:20)
at transformScriptBeforeCompilation (/opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:447:23)
at prepareScript (/opt/iobroker/node_modules/iobroker.javascript/main.js:2068:39)
at /opt/iobroker/node_modules/iobroker.javascript/main.js:2153:17
at Immediate. (/opt/iobroker/node_modules/iobroker.javascript/main.js:1628:17)
at processImmediate (node:internal/timers:478:21)
2023-06-12 07:42:13.089 - error: javascript.0 (26815) ts.createNodeArray is not a function
2023-06-12 07:42:13.135 - info: javascript.0 (26815) terminating
2023-06-12 07:42:13.137 - warn: javascript.0 (26815) Terminated (UNCAUGHT_EXCEPTION): Without reason
2023-06-12 07:42:13.673 - info: javascript.0 (26815) Start javascript script.js.Test.Sayit
2023-06-12 07:42:13.737 - info: javascript.0 (26815) script.js.Test.Sayit: lautstaerke:95;text: Hallo Bruno Hallo Marlon
2023-06-12 07:42:13.739 - info: javascript.0 (26815) script.js.Test.Sayit: registered 0 subscriptions, 0 schedules, 0 messages, 0 logs and 0 file subscriptions

**Comment by klein0r** [comment]
@Harry-58 Kannst Du mal meine Version testen ob damit alles läuft?

**Comment by Harry-58** [comment]
@klein0r 
Mit deiner Version laufen auch die ts-Scripte wieder, ohne dass der Adapter crashed.

**Comment by agross** [comment]
@klein0r Ich hab's mit deiner master-Version probiert. Geht leider nicht.

Das ist das betreffende (als erstes kompilierte) Script: https://github.com/agross/iobroker-scripts/blob/master/home/global/Shared/Device.ts

Lokal kompiliert es problemlos nachdem ich die TS-dependency auf `~5.1.3` anhebe.

<details>

```
2023-06-12 09:23:37.215  - info: javascript.0 (629891) script.js.global.Shared.Device: compiling TypeScript source...
2023-06-12 09:23:37.399  - error: javascript.0 (629891) Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2023-06-12 09:23:37.400  - error: javascript.0 (629891) unhandled promise rejection: Debug Failure. Invalid cast. The supplied value [object Object] did not pass the test 'isIdentifier'.
2023-06-12 09:23:37.401  - error: javascript.0 (629891) Error: Debug Failure. Invalid cast. The supplied value [object Object] did not pass the test 'isIdentifier'.
    at cast (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:1170:18)
    at pipelineEmitWithHintWorker (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111076:31)
    at pipelineEmitWithHint (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111061:9)
    at pipelineEmitWithComments (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:114845:7)
    at pipelineEmit (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111010:7)
    at emitIdentifierName (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:110984:7)
    at emitClassDeclarationOrExpression (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:112891:9)
    at emitClassDeclaration (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:112873:7)
    at pipelineEmitWithHintWorker (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111243:20)
    at pipelineEmitWithHint (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111061:9)
    at pipelineEmitWithComments (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:114845:7)
    at pipelineEmit (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111010:7)
    at emit (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:110978:7)
    at emitListItemWithParenthesizerRule (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:115291:5)
    at emitNodeListItems (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:114002:9)
    at emitNodeList (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:113941:9)
    at emitList (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:113899:7)
    at emitSourceFileWorker (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:113579:7)
    at emitBodyWithDetachedComments (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:114951:9)
    at emitSourceFile (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:113507:9)
    at pipelineEmitWithHintWorker (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111074:16)
    at pipelineEmitWithHint (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111061:9)
    at pipelineEmit (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:111010:7)
    at print (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:110923:7)
    at writeFile2 (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:110907:7)
    at printFile (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:110747:7)
    at Object.printNode (/opt/iobroker/node_modules/iobroker.javascript/node_modules/typescript/lib/typescript.js:110724:18)
    at transformScriptBeforeCompilation (/opt/iobroker/node_modules/iobroker.javascript/lib/typescriptTools.js:447:31)
    at Immediate.<anonymous> (/opt/iobroker/node_modules/iobroker.javascript/main.js:1123:67)
    at processImmediate (node:internal/timers:478:21)
2023-06-12 09:23:37.402  - error: javascript.0 (629891) Debug Failure. Invalid cast. The supplied value [object Object] did not pass the test 'isIdentifier'.
2023-06-12 09:23:37.448  - info: javascript.0 (629891) terminating
2023-06-12 09:23:37.451  - warn: javascript.0 (629891) Terminated (UNCAUGHT_EXCEPTION): Without reason
```

</details>

**Comment by klein0r** [comment]
> Lokal kompiliert es problemlos nachdem ich die TS-dependency auf `~5.1.3` anhebe.

Aber genau darauf steht doch die Dependency bei 7.0.5 gerade?

https://github.com/ioBroker/ioBroker.javascript/blob/ac3cd090c9770e8717eefcc329f36afa1dbf1b26/package.json#L45

**Comment by agross** [comment]
Ja, ich nutze für den lokalen `npm run build` immer dieselben Deps wie die die der Adapter vorgibt. Einer der Maintainer hier hat aber für TS und "globale" Scripte aber Magic in den Adapter eingefügt. Das habe ich lokal nur als Näherung mit Projektfiles abbilden können. Von daher ist der lokale Build meiner Skripte nicht 100% derselbe wie der der im Adapter ausgeführt wird.

Reproduzierbar ist, dass iobroker.javascript 7.0.5 nicht mit meinen Skripten klarkommt, 7.0.3 aber schon. Ob's an dem Einsatz globaler Skripte liegt oder am konkrekten Skript vermag ich nicht zu sagen. Ob andere im Thread hier globale Skripte im Einsatz haben wäre interessant zu wissen.

**Comment by klein0r** [comment]
> Reproduzierbar ist, dass iobroker.javascript 7.0.5 nicht mit meinen Skripten klarkommt

Über den Punkt sind wir ja schon weg. Dank der TS-Dependency 5.x kommt der Adapter mit gar keinem TypeScript mehr klar, da dort manche Funktionen einfach nicht mehr verfügbar sind. Daher ja mein Fix. Damit funktioniert es laut @Harry-58 ja wieder.

> Ob's an dem Einsatz globaler Skripte liegt

Ich habe nicht alles getestet. Globale Scripts z.B. nicht. Daher ja die Bitte das alles durchzuprobieren.

**Comment by agross** [comment]
> Daher ja die Bitte das alles durchzuprobieren.

Was ich getan habe und hier mitgeteilt habe was nicht funktioniert.

Jetzt wurde das Ticket durch @GermanBluefox geschlossen. Interessant.

**Comment by klein0r** [comment]
> Jetzt wurde das Ticket durch @GermanBluefox geschlossen. Interessant.

Das wurde geschlossen, weil mein PR gemerged wurde.

**Comment by agross** [comment]
> Daher ja mein Fix. Damit funktioniert es laut @Harry-58 ja wieder.

Vielleicht hatte es das nicht präzise genug ausgedrückt: Ich hatte deinen Fix/PR installiert, um zu testen ob 7.0.5 + dein Fix funktioniert.
